### PR TITLE
Metronome Effect + Effect Sync refectoring

### PR DESCRIPTION
**Pull request by daschuer**
While debugging the effect sync issues, I have decide that it is handy to have the Metronome effect requested in https://bugs.launchpad.net/mixxx/+bug/1019822 ready. 

You can either make the beat grit hear able, or select an independent BPM value.  

The issue was that the sync infos are source track sample rate dependent. Now, the values are passed time scaled. I have also decided to remove all unused stuff form the GroupFeatureState. This can be re-add on demand in a way that it suits well. For now it is better to save some CPU and not predict a future demand.

### CODE REVIEW

File: src/effects/native/metronomeclick.h [hunk 1]
```diff
@@ -0,0 +1,1034 @@
+#ifndef METRONOMECLICK_H
+#define METRONOMECLICK_H
+
+#include "util/types.h"
+
+const unsigned int kClickSize = 1024;
+const CSAMPLE kClick[kClickSize] = {
```
Comment (Be-ing) [inline src/effects/native/metronomeclick.h]: How did you generate these values?

File: src/effects/native/metronomeeffect.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    MetronomeEffect(EngineEffect* pEffect, const EffectManifest& manifest);
+    virtual ~MetronomeEffect();
+
+    static QString getId();
+    static EffectManifest getManifest();
+
+    // See effectprocessor.h
+    void processChannel(const ChannelHandle& handle,
+                        MetronomeGroupState* pState,
+                      const CSAMPLE* pInput, CSAMPLE* pOutput,
```
Comment (Be-ing) [inline src/effects/native/metronomeeffect.h]: whitespace is misaligned

File: src/engine/bpmcontrol.cpp [hunk 3]
```diff
@@ -874,14 +862,8 @@ void BpmControl::collectFeatures(GroupFeatureState* pGroupFeatures) const {
...
-        pGroupFeatures->has_prev_beat = true;
-        pGroupFeatures->prev_beat = dThisPrevBeat;
-
-        pGroupFeatures->has_next_beat = true;
-        pGroupFeatures->next_beat = dThisNextBeat;
-
-        pGroupFeatures->has_beat_length = true;
-        pGroupFeatures->beat_length = dThisBeatLength;
+        pGroupFeatures->has_beat_length_sec = true;
+        pGroupFeatures->beat_length_sec = dThisBeatLength / m_pTrack->getSampleRate() / 2;
```
Comment (Be-ing) [inline src/engine/bpmcontrol.cpp]: This assumes a stereo signal. Does it assume that the file is stereo or that the signal passing through the engine is stereo?

**Review by Be-ing — CHANGES_REQUESTED** [review]
Thanks for taking care of this. I like the idea of normalizing the value passed through the GroupFeatureState to a unit of time. I agree there's no need to waste a little CPU on collecting attributes for the GroupFeatureState that no effects use. Adding them back will be easy if there is a use case for them.

The metronome effect should mix the input signal with the click, not just output the click.

While the metronome effect works with its SYNC button parameter enabled and changing the engine's sample rate, when the SYNC button parameter is disabled, the speed of the click changes when the engine's sample rate is changed.

File: src/effects/native/metronomeclick.h (hunk 1 above)
Comment (daschuer) [reply src/effects/native/metronomeclick.h]: This is a recording from my daughter's metronome.

Comment (daschuer) [reply src/effects/native/metronomeclick.h]: I have used Audacity to export the samples as text.

File: src/engine/bpmcontrol.cpp (hunk 3 above)
Comment (daschuer) [reply src/engine/bpmcontrol.cpp]: dThisBeatLength is the count of stereo samples. I will add a comment.

**Comment by daschuer** [comment]
> The metronome effect should mix the input signal with the click, not just output the click.

I had considered this as well. But IMHO it was more hand to use the dry wet knob to change between only track and only metronome. 

For me this effect is more a technical helper used for PFL and not played out to the audience.
It can be used to set a beatgrid for dificult tracks. And probably for beatmatching using ears.
For both cases it could be handy to have a metronome only mode.

What is your use case?

File: src/effects/native/metronomeclick.h (hunk 1 above)
Comment (Be-ing) [reply src/effects/native/metronomeclick.h]: Please add a comment noting this.

**Comment by Be-ing** [comment]
> I had considered this as well. But IMHO it was more hand to use the dry wet knob to change between only track and only metronome.

The trouble with this, as the effect is currently implemented, is that turning the dry/wet knob towards the wet side reduces the volume of the dry signal in addition to bringing in the metronome sound. That would not happen if the effect mixed the input signal with the metronome sound.

> It can be used to set a beatgrid for dificult tracks.

This is the only practical use case I can think of for this effect. I had thought it might be helpful for playing an instrument in time, but in that case listening to the playing track would likely make more sense.

This effect could also be helpful for developing a way to quantize beat synchronized effects.

**Comment by daschuer** [comment]
> The trouble with this, as the effect is currently implemented, is that turning the dry/wet knob towards the wet side reduces the volume of the dry signal in addition to bringing in the metronome sound. That would not happen if the effect mixed the input signal with the metronome sound.

Yes, that might be an issue. For me it is handy to have pure track <-> pure metronome on a single knob and I can effort the reduced gain in the center position. The same can be achieved using gain and mix in a send version. Should I change it to this to the later?

File: src/effects/native/metronomeclick.h [hunk 4]
```diff
@@ -3,6 +3,9 @@
 
 #include "util/types.h"
 
+// This is a new recording from a real metronom
```
Comment (Be-ing) [inline src/effects/native/metronomeclick.h]: typo: metronome

File: src/engine/bpmcontrol.cpp [hunk 5]
```diff
@@ -863,7 +863,8 @@ void BpmControl::collectFeatures(GroupFeatureState* pGroupFeatures) const {
                        dThisPrevBeat, dThisNextBeat,
                        &dThisBeatLength, &dThisBeatFraction)) {
         pGroupFeatures->has_beat_length_sec = true;
-        pGroupFeatures->beat_length_sec = dThisBeatLength / m_pTrack->getSampleRate() / 2;
+		// Note: dThisBeatLength is fractional frames count * 2 (stereo samples)
```
Comment (Be-ing) [inline src/engine/bpmcontrol.cpp]: align comment indentation with the code

**Review by Be-ing — CHANGES_REQUESTED** [review]
> Yes, that might be an issue. For me it is handy to have pure track <-> pure metronome on a single knob and I can effort the reduced gain in the center position. The same can be achieved using gain and mix in a send version. Should I change it to this to the later?

Yes, please change it.

Are you able to reproduce the issue I encountered when changing the engine sample rate with the sync parameter disabled?

**Comment by daschuer** [comment]
> While the metronome effect works with its SYNC button parameter enabled and changing the engine's sample rate, when the SYNC button parameter is disabled, the speed of the click changes when the engine's sample rate is changed.

You mean, the pitch of the click sound changes right? That is true and I know this. 
This can be fixed either by adding a pitch changing operation or adding pre-calculated clicks for all sample rates. 
I did not do this, because a pitched click is still a nice click, and IMHO the pitch change can be ignored in favour of a simple implementation.

**Comment by Be-ing** [comment]
Yes, I was referring to the pitch change. I do not think it sounds good at 96 kHz how it is. I think using a precalculated set of samples for each sample rate would be good.

**Comment by daschuer** [comment]
I am in doubt if it is worth to spend an other 2 kB of memory food print for this effect, which is unlikely played to the audience. 
Do you think it is?

**Comment by Be-ing** [comment]
2 KB is a trivial amount of memory in the context of an application that requires hundreds of MB.

**Comment by daschuer** [comment]
Done! Now we consume 6696 bytes more.

File: src/effects/native/metronomeeffect.cpp [hunk 6]
```diff
@@ -77,15 +77,25 @@ void MetronomeEffect::processChannel(const ChannelHandle& handle, MetronomeGroup
         return;
     }
 
+    unsigned int clickSize = kClickSize44100;
+    const CSAMPLE* click = kClick44100;
+    if (sampleRate <= 48000) {
```
Comment (Be-ing) [inline src/effects/native/metronomeeffect.cpp]: This will be true when the sample rate is 44100 Hz.

**Comment by daschuer** [comment]
Ups, fixed.

**Comment by Be-ing** [comment]
Okay, LGTM! Thanks for fixing this.

**Comment by daschuer** [comment]
Thank you for review.

File: src/engine/bpmcontrol.cpp [hunk 7]
```diff
@@ -874,14 +862,9 @@ void BpmControl::collectFeatures(GroupFeatureState* pGroupFeatures) const {
...
-        pGroupFeatures->prev_beat = dThisPrevBeat;
-
-        pGroupFeatures->has_next_beat = true;
-        pGroupFeatures->next_beat = dThisNextBeat;
-
-        pGroupFeatures->has_beat_length = true;
-        pGroupFeatures->beat_length = dThisBeatLength;
+        pGroupFeatures->has_beat_length_sec = true;
+        // Note: dThisBeatLength is fractional frames count * 2 (stereo samples)  
+        pGroupFeatures->beat_length_sec = dThisBeatLength / m_pTrack->getSampleRate() / 2;
```
Comment (Be-ing) [inline src/engine/bpmcontrol.cpp]: I am doubting that this division by 2 should be here. It requires every effect to do this inverse operation (multiply by 2).

Comment (Be-ing) [reply src/engine/bpmcontrol.cpp]: On second thought it may be better to leave it as is and move effect code towards working with frames.

Comment (daschuer) [reply src/engine/bpmcontrol.cpp]: Yes, we should move to frames whenever no external interface change is required.
Playpositions are stored in stereo samples in our database, that can not be changed.

**Comment by ronso0** [comment]
A synced metronom and a 1-beat loop, then Echo, make a nice toy for really weird transitions..
Thanks a lot

**Comment by esbrandt** [comment]
@daschuer 
Trying the effect for the first time, just asking if the following behavior is intentional:
* Assign a deck to FX rack and activate metronome effect bpm-synced
* Route the master mix trough the effect chain ( by clicking the MASTER button in the rack)
* Hear a double-tap metronome click, the first synced to the deck BPM, the second to the Master BPM.

Since we do not expose the master BPM (not ready yet IIRC), can`t we discard the click on Master for now?

**Comment by daschuer** [comment]
Yes, that could make sense.
Original this effect was a developers tool, has it any user value for you?

**Comment by esbrandt** [comment]
> Original this effect was a developers tool, has it any user value for you?

Since our waveform zoom is not that great, and a lot of the available height on the RGB waveform is wasted, i found aligning beatgrids by visual guidance alone is tricky.
The tick helps me to align beatgrids by ear.

**Comment by daschuer** [comment]
I was just about to disable the effect for master, but looking at it again I think now you just hit a general problem for all synced effects. If you apply them twice for a deck and the master, they do just what they should, add the effect twice. This is probably always an unwanted state, but I think it should be not solved in a single effect.

**Comment by Be-ing** [comment]
Why would that be an issue? I think it is upon users to understand that would apply the effect twice. If Mixxx went out of the way to compensate for that situation, I think it could be unexpected that the effect is _not_ applied twice. Implementing that would require adding special case hacks to the effects system which IMO would be an unnecessary maintenance burden.

**Pull request merged**